fish-rust/src/builtins/path.rs

```diff
@@ -0,0 +1,987 @@
+        /// A directory
+        const DIR = 1 << 1;
+        /// A regular file
+        const FILE = 1 << 2;
+        /// A link
+        const LINK = 1 << 3;
+        /// A character device
+        const CHAR = 1 << 4;
+        /// A fifo
+        const FIFO = 1 << 5;
+        /// A socket
+        const SOCK = 1 << 6;
+    }
+}
+
+impl TryFrom<&wstr> for TypeFlags {
+    type Error = ();
+
+    fn try_from(value: &wstr) -> Result<Self, Self::Error> {
+        let flag = match value {
+            t if t == "file" => Self::FILE,
+            t if t == "dir" => Self::DIR,
+            t if t == "block" => Self::BLOCK,
+            t if t == "char" => Self::CHAR,
+            t if t == "fifo" => Self::FIFO,
+            t if t == "socket" => Self::SOCK,
+            t if t == "link" => Self::LINK,
+            _ => return Err(()),
+        };
+
+        Ok(flag)
+    }
+}
+
+bitflags! {
+    #[derive(Default)]
+    pub struct PermFlags: u32 {
+        const READ = 1 << 0;
+        const WRITE = 1 << 1;
+        const EXEC = 1 << 2;
+        const SUID = 1 << 3;
+        const SGID = 1 << 4;
+        const USER = 1 << 5;
+        const GROUP = 1 << 6;
+    }
+}
+
+impl PermFlags {
+    fn is_special(self) -> bool {
+        self.intersects(Self::SUID | Self::SGID | Self::USER | Self::GROUP)
+    }
+}
+
+impl TryFrom<&wstr> for PermFlags {
+    type Error = ();
+
+    fn try_from(value: &wstr) -> Result<Self, Self::Error> {
+        let flag = match value {
+            t if t == "read" => Self::READ,
+            t if t == "write" => Self::WRITE,
+            t if t == "exec" => Self::EXEC,
+            t if t == "suid" => Self::SUID,
+            t if t == "sgid" => Self::SGID,
+            t if t == "user" => Self::USER,
+            t if t == "group" => Self::GROUP,
+            _ => return Err(()),
+        };
+
+        Ok(flag)
+    }
+}
+
+/// This is used by the subcommands to communicate with the option parser which flags are
+/// valid and get the result of parsing the command for flags.
+#[derive(Default)]
+struct Options<'args> {
+    null_in: bool,
+    null_out: bool,
+    quiet: bool,
+
+    invert_valid: bool,
+    invert: bool,
+
+    relative_valid: bool,
+    relative: bool,
+
+    reverse_valid: bool,
+    reverse: bool,
+
+    unique_valid: bool,
+    unique: bool,
+
+    key: Option<&'args wstr>,
+
+    types_valid: bool,
+    types: Option<TypeFlags>,
+
+    perms_valid: bool,
+    perms: Option<PermFlags>,
+
+    arg1: Option<&'args wstr>,
+}
+
+fn path_out(streams: &mut io_streams_t, opts: &Options<'_>, s: &wstr) {
+    if !opts.quiet {
+        if !opts.null_out {
+            streams
+                .out
+                .append_with_separation(s, separation_type_t::explicitly, true);
+        } else {
+            // Note the char - if this was a string instead we'd add
+            // a string of length 0, i.e. nothing
+            streams.out.append(s.to_owned() + L!("\0"));
+        }
+    }
+}
```

Comment:
I did some thinking about making this generic to accept both `WString`, `&wstr`, and `Cow<'_, wstr>`, to avoid allocating once or twice in `s.to_owned() + L!("\0")`. The preliminary results are:

- `Cow<'a, str>` implements `Add<&'a str>`, which specialises where the cow's str is empty/ensures only one allocation happens if needed, but we don't use `str`.
    - Even if we could, we would need to make it into a Cow at the call-site
    - I contemplated making the Add in Cow generic, but that would lead to the same problem of up to two allocs
- We could make `path_out` a macro and use [`spez!`](https://github.com/m-ou-se/spez), this ends up with likely the least amount of code.
- Have our own `ToPathOut` trait  and use that with impls for Cow/WString/&wstr, a less hacky  solution
- Alternative: Alloc a `WString` with correct capacity and add `s` and `\0` to it. Results in always one alloc, which I think reflects the C++ version best?

For context, this was the C++ version:

```cpp
static void path_out(io_streams_t &streams, const options_t &opts, const wcstring &str) {
    if (!opts.quiet) {
        if (!opts.null_out) {
            streams.out.append_with_separation(str, separation_type_t::explicitly);
        } else {
            // Note the char - if this was a string instead we'd add
            // a string of length 0, i.e. nothing
            streams.out.append(str + L'\0');
        }
    }
}
```

Reply:
What you implemented is perfect: simple and easy. In every benchmark I've seen, fish performance is bound by system calls, in particular `fork`, `posix_spawn`, and (bizarrely) `close` for closing pipes associated with command redirections.